text/0000-gen-fn.md

```diff
@@ -0,0 +1,785 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## New keyword
+
+Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, etc. named `gen` must be renamed.
+
+## Returning/finishing an iterator
+
+`gen` blocks must diverge or return the unit type.  Specifically, the trailing expression must be of the unit or `!` type, and any `return` statements in the block must either be given no argument at all or given an argument of the unit or `!` type.
+
+### Diverging iterators
+
+For example, a `gen` block that produces the infinite sequence `0, 1, 0, 1, 0, 1, ...`, will never return `None`
+from `next`, and only drop its captured data when the iterator is dropped:
+
+```rust
+gen {
+    loop {
+        yield 0;
+        yield 1;
+    }
+}
+```
+
+If a `gen` block panics, the behavior is very similar to `return`, except that `next` unwinds instead of returning `None`.
+
+## Error handling
+
+Within `gen` blocks, the `?` operator desugars as follows.  When its
+argument returns a value indicating "do not short circuit"
+(e.g. `Option::Some(..)`, `Result::Ok(..)`, `ControlFlow::Continue(..)`), that
+value becomes the result of the expression as usual.  When its argument
+returns a value indicating that short-circuiting is desired
+(e.g. `Option::None`, `Result::Err(..)`, `ControlFlow::Break(..)`), the value
+is first yielded (after being converted by `From::from` as usual), then the
+block returns immediately.
+
+Even when `?` is used within a `gen` block, the block must return a
+value of type unit or `!`.  That is, it does not return a value of `Some(..)`,
+`Ok(..)`, or `Continue(..)` as other such blocks might.
+
+However, note that when `?` is used within a `gen` block, all `yield`
+statements will need to be given an argument of a compatible type.  For
+example, if `None?` is used in an expression, then all `yield` statements will
+need to be given arguments of type `Option`.
+
+## Fusing
+
+Like `Generators`, `Iterator`s produced by `gen` panic when invoked again after they have returned `None` once.
+This will probably be fixed by special casing the generator impl if `Generator::Return = ()`, as we can trivially
+produce infinite values of the unit type.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+## New keyword
+
+In the 2024 edition we reserve `gen` as a keyword. Previous editions will use `k#gen` to get the same features.
+
+## Error handling
+
+`foo?` in `gen` blocks will stop iteration after the first error by desugaring to:
+
+```rust
+match foo.branch() {
+    ControlFlow::Break(err) => {
+        yield R::from_residual(err);
+        return;
+    },
+    ControlFlow::Continue(val) => val,
+}
+```
+
+This is the same behaviour that `collect::<Result<_, _>>()` performs
+on iterators over `Result`s.
```

Comment:
I don't get this comparison, and I find the `?` extremely weird. Could you elaborate on the example?

`collect` turns an iterator of `Result` into a single `Result<Vec<_>>`, using the first error. How does that have anything to do with what generators do under the proposed semantics?

Reply:
The similarity is that both stop at the first error. But agreed that there's a difference in the types used; with `?` as described here you'd still have an iterator over a Result.

Reply:
Okay, so saying it's the "same" behavior is way off then IMO. It's somewhat similar but still rather different. The resulting behavior is an iterator yielding `Result` that promises to not yield more than one `Err`.

Reply:
Which really fits with `Coroutine<(), Yield = T, Return = Result<!, E>>` better. It might make sense to not support `?` inside a `Coroutine<(), Yield = Result<T, E>, Return = ()>` based generator and instead require usage of `try` blocks.

Reply:
This comparison stood out as unclear to me also, so we've now removed it from the document.